File: 0016-3Sum-Closest/lc0016.cc
// C++ solution.

// Two pointers.
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        if (nums.size() < 3) { return 0; }
        int ans = 0;
        int min_diff = numeric_limits<int>::max();

        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size()-2; i++) {
            if (i > 0 && nums[i] == nums[i-1]) { continue; }
            int left = i + 1;
            int right = nums.size() - 1;
            while (left < right) {
                int sum_three = nums[i] + nums[left] + nums[right];
                int diff = abs(sum_three - target);
                if (diff == 0) {
                    return sum_three;
                }
                if (diff < min_diff) {
                    min_diff = diff;
                    ans = sum_three;
                }
                if (sum_three > target) {
                    right--;
                } else {
                    left++;
                }
            }
        }
        return ans;
    }
};

File: 0016-3Sum-Closest/lc0016.rs
// Rust solution.

impl Solution {
    pub fn three_sum_closest(nums: Vec<i32>, target: i32) -> i32 {
        let mut nums = nums;
        let sz = nums.len();
        let (mut bestMatch, mut minDiff): (i32, i32) = (0, std::i32::MAX);
        if (sz < 3) { return -1; }
        else { nums.sort(); }

        for i in 0..(sz-2) {
            if (i > 0 && nums[i] == nums[i-1]) { continue; }
            let (mut left, mut right) = (i+1, sz-1);
            while (left < right) {
                let sum = nums[i] + nums[left] + nums[right];
                let diff = (sum-target).abs();
                if (diff == 0) { return target; }
                if (diff < minDiff) {
                    minDiff = diff;
                    bestMatch = sum;
                }
                if (sum > target) {
                    right -= 1;
                } else {
                    left += 1;
                }
            }
        }
        bestMatch
    }
}
